File: dart_core/lib/src/retry.dart
/// A predicate function that determines whether an error is retryable.
///
/// Returns `true` if the given [error] should trigger a retry attempt.
typedef RetryPredicate = bool Function(Object error);

/// Configuration for the retry mechanism.
///
/// Controls how many times a failed operation is retried and the delay
/// between attempts using exponential backoff.
///
/// Example:
/// ```dart
/// final config = RetryConfig(
///   maxAttempts: 5,
///   baseDelay: Duration(milliseconds: 500),
///   maxDelay: Duration(seconds: 30),
///   retryOn: (error) => error is NetworkException,
/// );
/// final result = await withRetry(() => apiCall(), config: config);
/// ```
class RetryConfig {
  /// The maximum number of total attempts (including the initial call).
  ///
  /// Defaults to `3`.
  final int maxAttempts;

  /// The base delay between retries.
  ///
  /// The actual delay doubles with each attempt (exponential backoff),
  /// up to [maxDelay]. Defaults to 1000ms.
  final Duration baseDelay;

  /// The maximum delay between retries.
  ///
  /// Caps the exponential backoff calculation. Defaults to 10000ms.
  final Duration maxDelay;

  /// A predicate that determines whether an error is retryable.
  ///
  /// Only errors for which this returns `true` will trigger a retry.
  /// Defaults to retrying errors whose runtime type contains `'Network'`.
  final RetryPredicate retryOn;

  /// Creates a [RetryConfig] with the given retry behavior.
  const RetryConfig({
    this.maxAttempts = 3,
    this.baseDelay = const Duration(milliseconds: 1000),
    this.maxDelay = const Duration(milliseconds: 10000),
    this.retryOn = _defaultRetryOn,
  });
}

/// Default retry predicate that retries errors whose type name contains `'Network'`.
bool _defaultRetryOn(Object error) {
  if (error is Exception) {
    // In Dart, network exceptions are the retryable ones
    return error.runtimeType.toString().contains('Network');
  }
  return false;
}

/// Executes [fn] with automatic retry and exponential backoff.
///
/// On failure, checks [RetryConfig.retryOn] to decide if the error is
/// retryable. If so, waits for `baseDelay * 2^(attempt-1)` (capped at
/// [RetryConfig.maxDelay]) before retrying.
///
/// [fn] is the async operation to execute.
/// [config] is the optional retry configuration. Uses [RetryConfig] defaults
/// if not provided.
///
/// Returns the result of [fn] on success.
///
/// Throws the last error from [fn] if all attempts fail or the error is
/// not retryable.
///
/// Example:
/// ```dart
/// final result = await withRetry(
///   () => payments.charge(params),
///   config: RetryConfig(maxAttempts: 3),
/// );
/// ```
Future<T> withRetry<T>(
  Future<T> Function() fn, {
  RetryConfig? config,
}) async {
  final cfg = config ?? const RetryConfig();
  Object? lastError;

  for (var attempt = 1; attempt <= cfg.maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;
      if (attempt >= cfg.maxAttempts || !cfg.retryOn(error)) {
        rethrow;
      }
      final delayMs = cfg.baseDelay.inMilliseconds * (1 << (attempt - 1));
      final cappedDelay = delayMs > cfg.maxDelay.inMilliseconds
          ? cfg.maxDelay
          : Duration(milliseconds: delayMs);
      await Future<void>.delayed(cappedDelay);
    }
  }
  throw lastError!;
}

File: ts-core/src/retry.ts
/**
 * Configuration for the retry wrapper.
 */
export interface RetryConfig {
  /** Maximum number of attempts (including the first call). Defaults to `3`. */
  maxAttempts?: number;

  /** Initial delay in milliseconds before the first retry. Defaults to `1000`. */
  baseDelayMs?: number;

  /** Maximum delay in milliseconds between retries. Defaults to `10000`. */
  maxDelayMs?: number;

  /**
   * Predicate that determines whether a failed attempt should be retried.
   * Receives the caught error and returns `true` to retry.
   * Defaults to retrying on `NETWORK_ERROR` and `TOKEN_FAILED` error codes.
   */
  retryOn?: (error: unknown) => boolean;
}

/**
 * Default retry predicate — retries on `NETWORK_ERROR` and `TOKEN_FAILED` error codes.
 */
const DEFAULT_RETRY_ON = (error: unknown): boolean => {
  if (error && typeof error === 'object' && 'code' in error) {
    const code = (error as { code: string }).code;
    return code === 'NETWORK_ERROR' || code === 'TOKEN_FAILED';
  }
  return false;
};

/**
 * Wraps an async function with retry logic using exponential backoff.
 *
 * On each failed attempt where `retryOn` returns `true`, the function is
 * retried after a delay of `baseDelayMs * 2^(attempt-1)`, capped at `maxDelayMs`.
 *
 * @typeParam T - Return type of the wrapped function.
 * @param fn - Async function to execute with retry.
 * @param config - Optional retry configuration.
 * @returns The result of `fn` on the first successful attempt.
 * @throws The last error if all attempts fail.
 *
 * @example
 * ```ts
 * const result = await withRetry(
 *   () => telebirr.payments.charge({ amount: '100', title: 'Test' }),
 *   { maxAttempts: 5, baseDelayMs: 500 },
 * );
 * ```
 */
export async function withRetry<T>(
  fn: () => Promise<T>,
  config?: RetryConfig,
): Promise<T> {
  const maxAttempts = config?.maxAttempts ?? 3;
  const baseDelayMs = config?.baseDelayMs ?? 1000;
  const maxDelayMs = config?.maxDelayMs ?? 10000;
  const retryOn = config?.retryOn ?? DEFAULT_RETRY_ON;

  let lastError: unknown;
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;
      if (attempt >= maxAttempts || !retryOn(error)) {
        throw error;
      }
      const delay = Math.min(baseDelayMs * 2 ** (attempt - 1), maxDelayMs);
      await new Promise((resolve) => setTimeout(resolve, delay));
    }
  }
  throw lastError;
}
